### ISSUE TITLE
`karafka server` not always stopping on TERM/INT signals

### ISSUE THREAD

**Issue by pat** [karafka/karafka-web#285]
## Expected behaviour

When running `bundle exec karafka server`, I expect that sending a INT or TERM signal will halt the server reasonably quickly.

## Actual behaviour

When running `bundle exec karafka server` _with karafka-web configured_, I find that sending TERM or INT signals (via `kill` or control-C) sometimes causes the process to say it's shutting down, but it doesn't actually stop. Sending a `KILL` signal does work reliably though.

It seems the problem is much, much more likely to happen if config.processing.interval is set to a higher number. Our code looks like the following, but I've also tried with 10,000 and have seem it happen then too.

```ruby
Karafka::Web.setup do |config|
  config.ui.sessions.secret = ENV.fetch("KARAFKA_WEB_SESSIONS_SECRET")

  # Reduce the logs/noise in development:
  config.processing.interval = 15 * 60 * 1000 if Rails.env.development?
end
```

I'm not sure if this is related to v0.8 or not. We're not yet running Kafka/karafka in production, and I've only noticed this problem as part of upgrading from v2.2.14/v0.7.10 to v2.3.3/v0.8.2 on my local machine today. We're only just starting to add consumers to our code, so maybe this issue existed in older releases it's just we hadn't been running `karafka server` much at all.

## Steps to reproduce the problem

I find with higher `config.processing.interval` values it's happening pretty much every time.

## Your setup details

```
Karafka version: 2.3.3
Ruby version: ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [arm64-darwin23]
Rdkafka version: 0.14.10
Consumer groups count: 2
Subscription groups count: 2
Workers count: 5
Application client id: [app]_development
Boot file: [Rails.root]/karafka.rb
Environment: development
License: LGPL-3.0
```

If any further details are useful, please do let me know!

**Comment by mensfeld** [comment]
Do you have any logs around this? In dev if your log level is `:debug` there should be a lot of polling reporting.

Does it hang beyond `shutdown_timeout`?

**Comment by mensfeld** [comment]
Ok I can reproduce.

**Comment by mensfeld** [comment]
It shutsdown correctly (not via forceful) but the question is, why does it take so much time.

**Comment by mensfeld** [comment]
As I see it, you trigger it yourself. The `processing.interval` is used for max wait time causing extensive wait time since the partition eof signaling will be available only in 2.4 and will be a default of karafka.

```
max_wait_time ::Karafka::Web.config.processing.interval / 2
```

**Comment by mensfeld** [comment]
This means, that karafka will run a super long poll of 15 minutes per your request before exiting the polling loop.

**Comment by mensfeld** [comment]
I don't think this is related to Karafka but your configuration that may cause this type of behavior. `enable.partition.eof` fast yielding will be available in 2.4 but even then in case of low-throughput topics with high wait time expectation karafka supervisor will wait for listeners to finish their operations.

I could consider early exit from the batch poll to improve this though for such cases.

**Comment by mensfeld** [comment]
ref: https://github.com/karafka/karafka/issues/1957

**Comment by mensfeld** [comment]
here you go: https://github.com/karafka/karafka/pull/1958

reduces shutdown time in your case from 15 minutes to 10 seconds tops.

**Comment by pat** [comment]
Thanks so much @mensfeld - it's just gone midnight here, but I'll give this a spin in the morning to confirm if it works well on my machine :)

**Comment by pat** [comment]
Finally got around to testing this, and I think I've found another part of the puzzle in what's related to the problem: we're also setting `config.shutdown_timeout` to 1500. If I remove that, using the latest commits to `karafka/karafka` and `karafka/karafka-web`, it seems to shut down reliably. But with that setting, it fails pretty regularly.

With these tests, I've kept `config.processing.interval` set to a very large number.

Here's some logging output, where I've used control-C at two separate points. I've also put some debugging lines within the client:

```ruby
@events_poller = Helpers::IntervalRunner.new do
  puts "POLLING EVENTS"
  events_poll
  # events poller returns nil when not running often enough, hence we don't use the
  # boolean to be explicit
  puts "BATCH POLL BREAKER: #{batch_poll_breaker.call}"
  batch_poll_breaker.call ? :run : :stop
end
```

```
Running in ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [arm64-darwin23]
Running Karafka 2.3.4 server
See LICENSE and the LGPL-3.0 for licensing details
[3c09b01e074b] Async producing of 1 messages to 1 topics took 12.58400011062622 ms
[e6fbfe37632d] Polled 1 messages in 12.382999897003174ms
[6c388d6e9465] Consume job for Karafka::Web::Processing::Consumer on karafka_consumers_reports/0 started
[3c09b01e074b] Async producing of 2 messages to 2 topics took 0.08599996566772461 ms
[6c388d6e9465] Consume job for Karafka::Web::Processing::Consumer on karafka_consumers_reports/0 finished in 110.56900000572205ms
POLLING EVENTS
BATCH POLL BREAKER: true
POLLING EVENTS
BATCH POLL BREAKER: true
^CReceived SIGINT system signal
Stopping Karafka server
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07899999618530273 ms
Forceful Karafka server stop
[3c09b01e074b] Async producing of 2 messages to 2 topics took 0.10299992561340332 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07599997520446777 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07600021362304688 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07699990272521973 ms
^CReceived SIGINT system signal
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07999992370605469 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07500004768371582 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07599997520446777 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.07999992370605469 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.0690000057220459 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.11700010299682617 ms
[3c09b01e074b] Async producing of 1 messages to 1 topics took 0.06999993324279785 ms
(and so on until I use `kill -KILL [pid]`)
```

For now I might avoid setting `config.shutdown_timeout` - the default seems fast enough - but perhaps you've some ideas of how these settings together aren't quite working.
